Git Bisect Combinator

adityagesh opened the pull request:
- Git bisect combinator clones code into a VM and loops the extended_phase in runbook
- Add KernelBuildMessage, which can be subscribed by notifiers
- Add "bisect" command in git tool


Sample Runbook:

`transformer:
  - type: azure_deploy
    name: bisect_vm
    requirement:
      azure:
        marketplace: $(marketplace_image)
        location: $(location)
      core_count: 2
    enabled: true
  - type: azure_deploy
    phase: expanded
    name: build_vm
    requirement:
      azure:
        marketplace: $(marketplace_image)
        location: $(location)
      core_count: 16
    enabled: true
  - type: kernel_installer
    phase: expanded
    connection:
      address: $(build_vm_address)
      private_key_file: $(admin_private_key_file)
    installer:
      type: source
      location:
        type: repo
        path: /mnt/code
        ref: $(ref)
        repo: "git://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git"
    raise_exception: False
  // Do not create vhd when build fails
  - type: azure_vhd
    enabled: $(kernel_installer_boot_success)
    phase: expanded
    resource_group_name: $(build_vm_resource_group_name)
    rename:
      azure_vhd_url: vhd
  // Do not run test when build fails
  - type: script
    phase: expanded
    scripts:
      - name: test_case_name
        variables:
          - test_case_name
          - kernel_installer_build_success
          - kernel_installer_boot_success
        script: "test_case_name if kernel_installer_build_success and kernel_installer_boot_success else ''"
  - type: azure_delete
    resource_group_name: $(build_vm_resource_group_name)
    phase: expanded_cleanup
  - type: azure_delete
    resource_group_name: $(bisect_vm_resource_group_name)
    phase: cleanup

combinator:
  - type: git_bisect
    repo: "git://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git"
    bad_commit: "HEAD"
    good_commit: "fe1de55167963a1c0ebe1579e37a8a41495f0a81"
    connection:
       address: $(bisect_vm_address)
       private_key_file: $(admin_private_key_file)

platform:
  - type: azure
    admin_private_key_file: $(admin_private_key_file)
    azure:
      subscription_id: $(subscription_id)
    requirement:
      azure:
        location: $(location)
        vhd: $(vhd)

testcase:
  - criteria:
      name: $(script_test_case_name)`

squirrelsc reviewed (commented):
Forgot to publish comments.

adityagesh committed:
Add Git Bisect Combinator

Git bisect combinator can be used to loop the runbook
until the bisect operations suceeds.

The example provided is to bisect failure in kernel tree
